### PR TITLE
Column Picker

### PR DESCRIPTION
**Pull request by abalogun-dev**
### Summary of changes 🔍 
User can add additional columns to the search results section of the Advanced Search modal

### Testing 🧪
Use the column picker dropdown and add some column and also deselect columns. Click on the column headers and check if it sorts appropriately.

https://msazure.visualstudio.com/One/_workitems/edit/15356066/
### Checklist ✔️
- [x] Linked associated issue (if present)
- [ ] Added relevant labels
- [ ] Requested developer reviews
- [ ] Request UI review from design / PM
- [ ] Verify all code checks are passing

### CODE REVIEW

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        return options;
+    };
+    return (
+        <Dropdown
+            ariaLabel="Custom dropdown example"
+            onRenderTitle={onRenderTitle}
+            options={setDropdownOptions()}
+            onChange={onChange}
+            multiSelect={true}
+            styles={{
```
Comment (pasanchMSFT) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx:109]: Could you add this to the styles file as a Subcomponent style?

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                key: 'Icon',
+                text: `${t('advancedSearch.propertyCount', {
+                    selectedPropsCount
+                })}`,
+                data: { icon: 'ColumnOptions' }
+            }
+        ];
+        return (
+            <div className={classNames.dropdownTitle}>
+                {placeholder[0].data && placeholder[0].data.icon && (
```
Comment (pasanchMSFT) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx:68]: If you just defined placeholder above, this is always going to be true. You don't need this check

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                setSelectedPropsCount(selectedPropsCount - 1);
+                deleteColumn(item.key as string);
+            }
+        }
+    };
+
+    const onRenderTitle = (): JSX.Element => {
+        const placeholder = [
+            {
+                key: 'Icon',
```
Comment (pasanchMSFT) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx:59]: Why do you need a key for placeholder?

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                    dropdownoptions['selected'] = true;
+                    dropdownoptions['disabled'] = true;
+                }
+            }
+        });
+        return options;
+    };
+    return (
+        <Dropdown
+            ariaLabel="Custom dropdown example"
```
Comment (pasanchMSFT) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx:104]: Loc this

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import { ComponentStory } from '@storybook/react';
+import ColumnPicker from './ColumnPicker';
+import { IColumnPickerProps } from './ColumnPicker.types';
+
+const wrapperStyle = { width: '100%', height: '600px', padding: 8 };
```
Comment (pasanchMSFT) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.stories.tsx:6]: Remove if unused

File: src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.types.ts [hunk 6]
```diff
@@ -33,8 +33,9 @@ export interface IAdvancedSearchResultDetailsListStyleProps {
 export interface IAdvancedSearchResultDetailsListStyles {
     root: IStyle;
     listHeader: IStyle;
+    listHeaderAndIcon: IStyle;
 
-    /**
+    /**s
```
Comment (pasanchMSFT) [inline src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.types.ts:38]: Remove s

File: src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx [hunk 7]
```diff
@@ -52,6 +53,9 @@ const AdvancedSearchResultDetailsList: React.FC<IAdvancedSearchResultDetailsList
     const listItems = twins.sort(
         sortAscendingOrDescending(sortKey, isSortedDescending)
     );
+    const [selectedColumnNames, setSelectedColumnNames] = useState<string[]>(
```
Comment (pasanchMSFT) [inline src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx:56]: Really like this use of keys not an array of the whole option object.

File: src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx [hunk 8]
```diff
@@ -199,11 +198,37 @@ const AdvancedSearchResultDetailsList: React.FC<IAdvancedSearchResultDetailsList
...
         <section className={classNames.root}>
-            <h3 className={classNames.listHeader}>
-                {t('advancedSearch.results', { twinCount })}
-            </h3>
+            <div className={classNames.listHeaderAndIcon}>
+                <h3 className={classNames.listHeader}>
+                    {t('advancedSearch.results', { twinCount })}
+                </h3>
+                <ColumnPicker
+                    searchedProperties={searchedProperties}
```
Comment (pasanchMSFT) [inline src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx:226]: ColumnPicker is a styled component, please add the className as a subcomponent style.

**Review by pasanchMSFT — APPROVED** [review]
Logic and implementation look good, just suggested a few things here and there.

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+                property.localeCompare('$metadata') != 0
+            ) {
+                const dropdownoptions = {
+                    key: property,
+                    text: property
+                };
+                options.push(dropdownoptions);
+                if (searchedProperties.includes(property)) {
+                    dropdownoptions['selected'] = true;
+                    dropdownoptions['disabled'] = true;
```
Comment (msnyder-msft) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx:94]: As we discussed, let's remove the disabled so they can still remove them.

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        return options;
+    };
+    return (
+        <Dropdown
+            ariaLabel={t('availableProperties')}
+            onRenderTitle={onRenderTitle}
+            options={setDropdownOptions()}
+            onChange={onChange}
+            multiSelect={true}
+            styles={{
```
Comment (msnyder-msft) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx:107]: can we pull these out to the styles file and use subComponentStyles please?

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+                root: { paddingRight: 28, width: 200 },
+                title: { alignItems: 'center' },
+                dropdownItems: { width: 200 }
```
Comment (msnyder-msft) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx:110]: what happens if we don't ahve the width on the dropdownItems?

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.types.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * SubComponent styles.
+     */
+    subComponentStyles?: IColumnPickerSubComponentStyles;
+}
+
+export interface IColumnPickerSubComponentStyles {
+    icon?: IIconStyles;
+    dropDown?: IDropdownStyles;
```
Comment (msnyder-msft) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.types.ts:38]: nit: casing `dropdown`

**Review by msnyder-msft — COMMENTED** [review]
Please fix the merge conflicts so we can get a storybook branch and then I'll test the UX.

**Review by pasanchMSFT — APPROVED** [review]

File: src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx [hunk 13]
```diff
@@ -201,11 +199,37 @@ const AdvancedSearchResultDetailsList: React.FC<IAdvancedSearchResultDetailsList
...
     return (
         <section className={classNames.root}>
-            <h3 className={classNames.listHeader}>
-                {t('advancedSearch.results', { twinCount })}
-            </h3>
+            <div className={classNames.listHeaderAndIcon}>
+                <h3 className={classNames.listHeader}>
+                    {t('advancedSearch.results', { twinCount })}
+                </h3>
+                <ColumnPicker
```
Comment (msnyder-msft) [inline src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx:226]: please hide the picker when there are no results
image

File: src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx [hunk 14]
```diff
@@ -201,11 +199,37 @@ const AdvancedSearchResultDetailsList: React.FC<IAdvancedSearchResultDetailsList
...
+    };
     return (
         <section className={classNames.root}>
-            <h3 className={classNames.listHeader}>
-                {t('advancedSearch.results', { twinCount })}
-            </h3>
+            <div className={classNames.listHeaderAndIcon}>
+                <h3 className={classNames.listHeader}>
+                    {t('advancedSearch.results', { twinCount })}
+                </h3>
```
Comment (msnyder-msft) [inline src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx:225]: it doens't default to the right list. I don't see the query columns getting added.
image

File: src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx [hunk 15]
```diff
@@ -201,11 +199,37 @@ const AdvancedSearchResultDetailsList: React.FC<IAdvancedSearchResultDetailsList
...
+        setSelectedColumnNames((currentValue) =>
+            currentValue.filter((col) => col !== columnToRemoveKey)
+        );
+    };
     return (
         <section className={classNames.root}>
-            <h3 className={classNames.listHeader}>
-                {t('advancedSearch.results', { twinCount })}
-            </h3>
+            <div className={classNames.listHeaderAndIcon}>
```
Comment (msnyder-msft) [inline src/Components/AdvancedSearch/Internal/AdvancedSearchResultDetailsList/AdvancedSearchResultDetailsList.tsx:222]: IMO. If you search, add columns, then search again, it should include the searched columns in the list of results while maintaining the ones you added. I don't mind if it readds search columns that were removed (though it'd be nice if it remembered that).

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    IColumnPickerStyleProps,
+    IColumnPickerStyles
+} from './ColumnPicker.types';
+import { getStyles } from './ColumnPicker.styles';
+import {
+    classNamesFunction,
+    useTheme,
+    styled,
+    Icon,
+    Dropdown,
```
Comment (msnyder-msft) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx:13]: when no columns are selected, we should still get a message. Maybe something like. "Select additional columns"
image

**Review by msnyder-msft — CHANGES_REQUESTED** [review]

File: src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+    addColumn,
+    deleteColumn,
+    styles
+}) => {
+    const classNames = getClassNames(styles, {
+        theme: useTheme()
+    });
+    const { t } = useTranslation();
+    const [selectedKeys, setSelectedKeys] = React.useState<string[]>([]);
+    const [selectedPropsCount, setSelectedPropsCount] = React.useState(
```
Comment (msnyder-msft) [inline src/Components/AdvancedSearch/Internal/ColumnPicker/ColumnPicker.tsx:35]: let's remove this state and simply reference `selectedKeys.length` so we don't need extra renders.

**Review by msnyder-msft — APPROVED** [review]
LGTM

**Pull request merged**